loopy/check.py

```diff
@@ -555,11 +634,81 @@ def map_if(self, expr, domain, insn_id):
+        super().map_call(expr, domain, insn_id)
+
+        import pymbolic.primitives as prim
+        from loopy.kernel.function_interface import (CallableKernel,
+                                                     get_kw_pos_association)
+        from loopy.symbolic import (guarded_aff_from_expr,
+                                    get_dependencies)
+        from loopy.diagnostic import ExpressionToAffineConversionError
+
+        if (isinstance(expr.function, ResolvedFunction)
+            and isinstance(self.callables_table[expr.function.name],
+                           CallableKernel)):
+
+            subkernel = self.callables_table[expr.function.name].subkernel
+
+            # The plan here is to add the constraints coming from the values
+            # args passed at a call-site as assumptions to the callee. To avoid
+            # variable-naming collision between the caller and callee, we
+            # prepend all the caller variable by "_lp_caller_".
+
+            # {{{ get bset from kwargs
+
+            kw_to_pos, _ = get_kw_pos_association(subkernel)
+
+            arg_id_to_arg = self.kernel.id_to_insn[insn_id].arg_id_to_arg()
+
+            kwargs = {k: _mark_variables_from_caller(arg_id_to_arg[kw_to_pos[k]])
+                      for k in subkernel.get_unwritten_value_args()}
+
+            kw_space = isl.Space.create_from_names(
+                subkernel.isl_context, set=[],
+                params=(get_dependencies(tuple(kwargs.values()))
+                        | set(kwargs.keys())))
+
+            extra_assumptions = isl.BasicSet.universe(kw_space).params()
+
+            for kw, arg in kwargs.items():
+                try:
+                    aff = guarded_aff_from_expr(extra_assumptions.space,
+                                                prim.Variable(kw) - arg)
+                except ExpressionToAffineConversionError:
+                    # arg expression not affine => don't add any constraints
+                    # corresponding to it
+                    continue
+
+                extra_assumptions = (extra_assumptions
+                                     .add_constraint(isl.Constraint
+                                                     .equality_from_aff(aff)))
 
-def _check_bounds_inner(kernel):
+            # take into account caller's assumptions
+            extra_assumptions = _align_and_intersect_with_caller_assumption(
+                extra_assumptions, self.kernel.assumptions)
+
+            # project out the assumptions on caller's variables as they don't
+            # bear any semantic meaning in the callee.
+            extra_assumptions = extra_assumptions.project_out_except(
+                types=[isl.dim_type.param],
+                names=subkernel.get_unwritten_value_args())
+
+            subkernel = subkernel.copy(
+                assumptions=_align_and_intersect(subkernel.assumptions,
+                                                 extra_assumptions))
```

Comment:
Technically, if an iname is passed as a function argument, the set it's in could also serve as assumptions for the callee. I'm OK with bumping that to an issue, too (and maybe adding a FIXME here). Just let me know.

Reply:
Thanks, considering inames was in my blindspot. I went with adding a fixme in 3325b82.

Reply:
#430